text/0000-snapshot-system.md

```diff
@@ -0,0 +1,207 @@
+- Spawn commands, with all the spawn parameters, the tick and a bitmask of target connections 
+
+![](0000-snapshot-system/snapshot.rfc.png)
+
+### The Snapshot Message 
+
+Machines will exchange Snapshot Messages. A Snapshot Message is nothing but a partial Snapshot (plus a header). Those messages will be used to synchronize the state and replicate events. Apart from containing less elements than Snapshots, the only difference between Snapshot Messages and Snapshot is that Snapshot messages don't need the bit mask for destination.
+
+Additionally, a Snapshot Message will include a header that contains:
+- a sequence number unique to this message
+- the tick it was sent on
+- acknowledgment for previous messages from this connection (in the form of a bitmask over the _n_ previous sequence numbers)
+
+![](0000-snapshot-system/snapshot.message.rfc.png)
+
+### Eventual Consistency
+
+The model we will use for synchronisation is eventual consistency. Any Snapshot Message might be lost, but retransmits(1) will have the information eventually arrive to its destination. In the case of packet loss, some variables still get to update.
+
+Dependencies between variable updates and spawning will be managed by the Snapshot System. You will only ever get updates to variables belonging to objects that were spawned. State information (NetworkVariable) is updated with latest available value, and intermediate values can be skipped. That happens, for example, if there's packet re-ordering in transit.
+
+Event information (spawns) is guaranteed to be applied only once, and in order. 
+
+(1) When we refer to retransmit, we mean that further Snapshot Messages will carry the same information again. Past Snapshot Messages will not be re-sent identically.
+
+### Enabling Snapshot
+
+Whether Snapshot is enabled or not, for NetworkVariables or Spawns, should have no visible impact to the user, apart for:
+- bandwidth improvements
+- delivery improvements
+- consistency between delivery of NetworkVariable updates and Spawning
+
+To enable or disable it, the following public variables will be offered in `NetworkConfig`:
+
+```
+public bool UseSnapshotVariable = false;
+bublic bool UseClassicSpawn = false;
+```
+
+The first one will control whether NetworkVariables are sent via the Snapshot System or not, and the second one whether the Spawns are sent via the Snapshot System.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+### Guiding principles
+
+The first guiding principle to introduce is that under normal operations **no** memory allocations are done at runtime, within the Snapshot System. Every buffer is pre-allocated when the Snapshot System is started. Provisioning is made for sending, for receiving, and for any internal processing.  
+
+Another guiding principle is that the Snapshot System doesn't care whether we have a star topology, a full mesh, or anything in between. Rather it is only concerned about:
```

Comment:
I think from a design perspective this is kind of correct. But in practice we would need to add a lot of additional logic to support multiple writer issues in a mesh topology.